bindings/go/blob/buffered_reader.go

```diff
@@ -0,0 +1,137 @@
+package blob
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/opencontainers/go-digest"
+)
+
+// NewEagerBufferedReader creates a new EagerBufferedReader instance.
+func NewEagerBufferedReader(r io.Reader) *EagerBufferedReader {
+	return &EagerBufferedReader{
+		reader:    r,
+		mediaType: "application/octet-stream",
+	}
+}
+
+// EagerBufferedReader is a reader that can calculate its digest and size
+// By eagerly loading the data, and proxying a buffer for it in memory.
+// This can lead to extraordinary memory usage for large files, so one should
+// be careful when using this without considering the potential size of the data.
+type EagerBufferedReader struct {
+	mu        sync.RWMutex
+	buf       *bytes.Buffer
+	digest    string
+	size      int64
+	loaded    bool
+	reader    io.Reader
+	mediaType string
+}
+
+var (
+	// By buffering the data, we can calculate the size and digest of the data
+	// from the buffer!
+	_ SizeAware   = &EagerBufferedReader{}
+	_ DigestAware = &EagerBufferedReader{}
+
+	// We can also set the media type of the buffer if needed
+	_ MediaTypeAware = &EagerBufferedReader{}
+)
+
+func (b *EagerBufferedReader) LoadEagerly() error {
+	if b.Loaded() {
+		return nil
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.buf == nil {
+		b.buf = &bytes.Buffer{}
+	}
+	if b.reader != nil {
+		b.reader = io.MultiReader(b.buf, b.reader)
+	}
+	dig, err := digest.FromReader(io.TeeReader(b.reader, b.buf))
+	if err != nil {
+		return err
+	}
+	b.digest = dig.String()
+	if newSize := int64(b.buf.Len()); newSize > b.size {
```

Comment:
```suggestion
	if newSize := int64(b.buf.Len()); newSize != b.size {
```

Why is `newSize` only set when it is bigger then the current `b.size`. What happens when the underlying blob changes and the blob is now smaller? Is such a scenario realistic?

Reply:
I think the size of a buffered reader should never decrease, only increase. This is because buffers are by definition allocation based and it would be strange to suddenly have a shrinking buffer. Thats also why I wrote

```
	// SetPrecalculatedSize sets the size of the blob if it is known in advance.
	// It MUST be safe for concurrent use, serializing as necessary.
	// It MUST be safe to call multiple times, with the same or greater different size.
	// Once called, HasPrecalculatedSize MUST return true.
	// If called with a size less than the current size, it MUST ignore this size.
	SetPrecalculatedSize(size int64)
```

Reply:
okay, that makes sense